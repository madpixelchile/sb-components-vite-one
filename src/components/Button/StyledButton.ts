import styled from "styled-components";

export const StyledButton = styled.button`

    
${({ theme }) => {
    const { colors } = theme;
    return `
            
          font-size: 16px;
          color: ${colors.primary};
          background-color: ${colors["btn-primary"]};
          border-radius: 6px;
          border: none;
          padding: 8px 24px;
          transition: all 300ms ease;
          cursor: pointer;

          border: solid 1px transparent;

          &:hover{
            color: white;
            background-color: ${colors["btn-primary-hover"]};
            transition: all 300ms ease;
          }
          &:active{
            background-color: #000;
          }
          &.as-link{
            background: none;
            padding: 0;
            display: inline-block;
            text-decoration: underline;
            color: ${colors['color-link']};
            transition: color 300ms ease;
            &:hover{
              color: ${colors['color-link-hover']};
              transition: color 0ms ease;
            }
          }

          &.outlined{
            background-color: transparent;
            border-color: ${colors["btn-primary"]};
            color: ${colors["btn-primary"]};

            &:hover{
              background-color: ${colors["btn-primary"]};
              color: ${colors.primary};
            }

            &:active{
              background-color: ${colors["btn-primary-hover"]};
              color: ${colors.primary};
            }

          }

          .loading-area{
            display: flex;
            align-items: center;
          }

        `
  }}

`;