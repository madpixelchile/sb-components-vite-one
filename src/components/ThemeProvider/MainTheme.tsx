

import { ThemeProvider } from 'styled-components';
import { StyledThemeProvider } from './StyledThemeProvider';
import { darkMode, lightMode } from '../../styles/theme';

export interface MainThemeProps {
  children?: React.ReactNode;
  theme?: string | {},
  [key: string]: any;
}

export const MainTheme = ({ children = null, theme = 'light' }:MainThemeProps) => {

  let selectedTheme: string | object = lightMode;

  if(typeof theme !== 'object'){
    if(theme === 'light'){
      //No object - string light - default lightModeAdded
      selectedTheme = lightMode;
    }else if(theme === 'dark'){
      //No object - string dark -  default darkModeAdded
      selectedTheme = darkMode;
    }
  }else if(typeof theme === 'object'){
    if(Object.keys(theme).length === 0){ //!PENDING Compare with the final BASE required properties
      //Empty object - Default lightMode added
      selectedTheme = lightMode;
    }else{
      //Full object - CustomMode or theme added
      selectedTheme = theme;
    }
  }

  return (
    <>
        <ThemeProvider theme={ selectedTheme }>
            <StyledThemeProvider/>
            { children }
        </ThemeProvider>
    </>
  )
}

export default  MainTheme;
