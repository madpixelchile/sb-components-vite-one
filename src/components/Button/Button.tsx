import React, { forwardRef, Ref } from "react";
import { StyledButton } from "./StyledButton";

export interface ButtonProps {
  children?: React.ReactNode;
  className?: string;
  variant?: string;
  [x: string]: any;
}

export const Button = forwardRef(({ children = '', className = '', variant = 'button',  ...props }:ButtonProps, ref: Ref<HTMLButtonElement>) => {

  let mainVariant = variant;

  if(variant === 'link'){
    mainVariant = 'as-link'
  }
  if(variant === 'outlined'){
    mainVariant = 'outlined'
  }

  return (
    <>
        <StyledButton
            className={`${ mainVariant } ${ className }`}
            ref={ref}
            {...props}
        >{ children }</StyledButton>
    </>
  )
})