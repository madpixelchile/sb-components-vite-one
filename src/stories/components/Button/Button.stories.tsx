import { Meta } from "@storybook/react";
import { Button, ButtonProps } from "../../../components/Button/Button";
import MainTheme from "../../../components/ThemeProvider/MainTheme";

const meta = {
    //Se recomienda generar un storybook a parte del proyecto final y con vite
    //El slash genera la visualización en carpetas y subniveles en la navegación principal
    title: 'Components/Button/Button',
    //El componente a incorporar en el muestrario
    component: Button,
    //Genera la documentación general del componente
    tags:[
        'autodocs'
    ],
    //Para centrar el componente en la mesa o cama (muestrario) de componente en pantalla
    parameters: {
        layout: 'centered'
    },
    //Para que en la documentación los tamaños salgan en un formato determinado
    argTypes:{
        className: { control: 'text' },
        variant: { control: 'select', options: ['','link', 'outlined'] },
        children: { control: 'text' }
    },

} satisfies Meta<typeof Button>;

export default meta;

//Ejemplo de componente con sus props específicas aplicadas
export const Basic = ({children = 'Click me', ...props}:ButtonProps)=>(
    <MainTheme>
        <Button { ...props }>{ children }</Button>
    </MainTheme>
    
)
