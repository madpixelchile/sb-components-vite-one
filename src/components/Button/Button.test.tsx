
import { render } from '@testing-library/react'; //screen screen.debug();
import { Button } from './Button';
import MainTheme from '../ThemeProvider/MainTheme';

describe('Testing Buttons', () => {

    test('Testing className & children', () => {

        const { getByText, container } = render(<MainTheme theme={'dark'}><Button className='custom-class'>Children</Button></MainTheme>);
        const childrenElement = getByText('Children');
        expect(childrenElement.innerHTML).toBe('Children');

        const classProp = container.getElementsByClassName('custom-class');
        expect(classProp.length).toBe(1)

        expect(container).toMatchSnapshot();

    })

    test('Testing variants link', () => {

        const { container } = render(<MainTheme theme={MainTheme}><Button variant={'link'} ></Button></MainTheme>);
        const classProp = container.getElementsByClassName('as-link');
        expect(classProp.length).toBe(1)

    })

    test('Testing variants outlined', () => {

        const { container } = render(<MainTheme theme={{}}><Button variant={'outlined'} ></Button></MainTheme>);
        const classProp = container.getElementsByClassName('outlined');
        expect(classProp.length).toBe(1)

    })

})