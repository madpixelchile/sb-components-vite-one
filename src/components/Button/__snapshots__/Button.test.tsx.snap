// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Buttons Testing className & children 1`] = `
<div>
  <button
    class="sc-bdnyFh gJedZo button custom-class"
  >
    Children
  </button>
</div>
`;
